### PR TITLE
Introduce a state archive column family

### PR DESCRIPTION
**Pull request by rphmeier**
Removes `ArchiveDB` from the codebase, replacing it with a variant of `OverlayRecentDB` which also inserts all values, canonical or otherwise, into a state archive column family which will be queried in archive mode as a fallback from the most recent state.

Also refines the `ethcore::db` module to encapsulate column-family related logic a little better.

This should lead to snapshots and sync on archive being faster. I am in the process of testing now.

### CODE REVIEW

**Comment by gavofyork** [comment]
marked in progress until it's properly tested...

**Comment by rphmeier** [comment]
this does result in much faster snapshots, as well as a lot of IO usage since it writes out every insertion to  two column families (which don't share memtables). The extra IO usage is pretty annoying, so I wonder if there would be a way to cut down on it a bit. Maybe the merge operator could help here?

@arkpar @nikvolf any ideas?

**Comment by arkpar** [comment]
The value should be inserted to the archive column only when it is deleted from the main column.

**Comment by rphmeier** [comment]
I initially tried that, but the problem is that the overlaydb journalling doesn't store values for deleted keys. So it'd also have to query the present state family to get the value before writing out to the other column family.

**Comment by rphmeier** [comment]
The current strategy is to write everything, including journalling, to the archive column family. Only canonical insertions and deletions are enacted to the state column family. Much of the queriable state remains in-memory, as is the case with standard overlayrecent pruning.

File: ethcore/src/db.rs [hunk 1]
```diff
@@ -14,26 +14,59 @@
...
+pub const COL_EXTRA: Option<u32> = Some(3);
+/// Column for Traces
+pub const COL_TRACE: Option<u32> = Some(4);
+/// Column for archived state.
+pub const COL_STATE_ARCHIVE: Option<u32> = Some(5);
+/// Number of columns in DB
+pub const NUM_COLUMNS: Option<u32> = Some(6);
+
+/// Create a new `JournalDB` trait object.
+pub fn make_journaldb(backing: Arc<Database>, algorithm: Algorithm) -> Box<JournalDB> {
```
Comment (NikVolf) [inline ethcore/src/db.rs]: `make_` as well as `get_` should be probably avoided for code without (logical) side effects

File: parity/migration.rs [hunk 2]
```diff
@@ -130,17 +130,31 @@ fn backup_database_path(path: &Path) -> PathBuf {
...
+pub fn legacy_migration_settings(compaction_profile: &CompactionProfile) -> MigrationConfig {
 	MigrationConfig {
 		batch_size: BATCH_SIZE,
 		compaction_profile: *compaction_profile,
+		default_cols: None,
+	}
+}
+
+/// Consolidated migration settings.
+pub fn consolidated_migration_settings(compaction_profile: &CompactionProfile) -> MigrationConfig {
```
Comment (tomusdrw) [inline parity/migration.rs]: Perhaps just `migration_settings`?

File: parity/migration.rs [hunk 3]
```diff
@@ -130,17 +130,31 @@ fn backup_database_path(path: &Path) -> PathBuf {
 	backup_path
 }
 
-/// Default migration settings.
-pub fn default_migration_settings(compaction_profile: &CompactionProfile) -> MigrationConfig {
+/// Legacy migration settings.
+pub fn legacy_migration_settings(compaction_profile: &CompactionProfile) -> MigrationConfig {
```
Comment (tomusdrw) [inline parity/migration.rs]: Can this be moved to `legacy` module?

File: util/src/migration/mod.rs [hunk 4]
```diff
@@ -272,7 +275,7 @@ impl Manager {
 		let migration = self.migrations.iter().find(|m| m.version() == version);
 		match migration {
 			Some(m) => m.columns(),
-			None => None
+			None => self.config.default_cols,
```
Comment (tomusdrw) [inline util/src/migration/mod.rs]: Why do we need this? Shouldn't there always be a migration with `version-1`?

File: ethcore/src/migrations/v10.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		ToV10 {
+			progress: Progress::default(),
+			pruning: pruning,
+		}
+	}
+}
+
+impl Migration for ToV10 {
+	fn version(&self) -> u32 { 10 }
+	fn columns(&self) -> Option<u32> { NUM_COLUMNS }
```
Comment (tomusdrw) [inline ethcore/src/migrations/v10.rs]: What about introducing `columns_before` `columns_after` here instead of using `default_cols` and `no_of_columns_at`?

File: util/src/journaldb/overlayrecentdb.rs [hunk 6]
```diff
@@ -99,34 +103,53 @@ const PADDING : [u8; 10] = [ 0u8; 10 ];
...
+		};
+
+		let reconstructed = Self::read_overlay(&self.backing, overlay_col);
 		let journal_overlay = self.journal_overlay.read();
 		*journal_overlay == reconstructed
 	}
 
 	fn payload(&self, key: &H256) -> Option<Bytes> {
-		self.backing.get(self.column, key).expect("Low-level database error. Some issue with your hard disk?")
+		match self.backing.get(self.column, key).expect("Low-level database error. Some issue with your hard disk?") {
```
Comment (tomusdrw) [inline util/src/journaldb/overlayrecentdb.rs]: By changing `self.mode` to `Option` (on this level (or perhaps `Into`)) you can get a bit more concise code:

```rust
let val = self.backing.get(self.column, key).expect("...");
val.or_else(|| {
  self.archive_column.map(|col| self.backing.get(col, key).expect(".."))
})
```
or
```rust
self.get_backing(self.column, key).or_else(|| {
  self.archive_column.map(|col| self.get_backing(col, key))
})
 ``

File: util/src/journaldb/overlayrecentdb.rs [hunk 7]
```diff
@@ -211,9 +234,22 @@ impl JournalDB for OverlayRecentDB {
...
+				Archive::On(archive_col) =>
+					self.backing.get_by_prefix(archive_col, &key[0..DB_PREFIX_LEN]).map(|b| b.to_vec())
+		})
 	}
 
 	fn commit(&mut self, batch: &DBTransaction, now: u64, id: &H256, end: Option<(u64, H256)>) -> Result<u32, UtilError> {
+		// default column to insert journalling and writes to.
+		// in archive mode, we put everything except canonical state in the
+		// archive column.
+		let default_col = match self.mode {
```
Comment (tomusdrw) [inline util/src/journaldb/overlayrecentdb.rs]: By converting to option: `self.archive_column.or(self.column)`

File: ethcore/src/migrations/v10.rs (hunk 5 above)
Comment (rphmeier) [inline ethcore/src/migrations/v10.rs]: that seems like a good idea; I just needed a quick fix :)

Comment (tomusdrw) [inline ethcore/src/migrations/v10.rs]: Sure. Can/Should be done in separate PR :)

**Comment by rphmeier** [comment]
Will close for now, save for the 1.4 series.
